modules/internal/ChapelArray.chpl

```diff
@@ -1733,15 +1744,80 @@ module ChapelArray {
        instance of ``val`` in the array, or if ``val`` is not found, a
        tuple containing ``false`` and an unspecified value is returned.
      */
-     @unstable "'Array.find' is unstable"
-     proc find(val: this.eltType): (bool, index(this.domain)) {
+     deprecated "The tuple-returning version of `.find() on arrays` is deprecated; to opt into the new index-returning version, recompile with `-suseNewArrayFind`"
+     proc find(val: this.eltType): (bool, index(this.domain)) where !useNewArrayFind {
       for i in this.domain {
         if this[i] == val then return (true, i);
       }
       var arbInd: index(this.domain);
       return (false, arbInd);
     }
 
+
+    /*
+
+      Search an array for ``val``, returning whether or not it is
+      found.  If the value is found, the index storing it is returned
+      in ``idx``.  If it is not found, the final value of ``idx`` is
+      unspecified.
+
+    */
+    proc find(val: eltType, ref idx: fullIdxType): bool {
+      // For the sparse case, start by seeing if the IRV is what we're
+      // looking for.  If so, iterate over the parent domain to look
+      // for the value or an index not represented in the array.  This
+      // assumes that we're likely to find a case of the IRV fast.
+      // Otherwise, fall through to the normal case to just search the
+      // explicitly represented values.
+      if this.isSparse() && val == this.IRV {
+        for i in this.domain._value.parentDom {
+          // If we find an index representing the IRV, return it; if
+          // it has an explicit value, it may still be the IRV, so
+          // check it.
+          if !this.domain.contains(i) || this[i] == val {
+            idx = i;
+            return true;
+          }
+        }
+      } else {
+        for i in this.domain {
```

Comment:
Do we want to use a parallel loop or reduction for this? I guess the idea is we'll come back and optimize this later at some point now that we've settled on the interface.

Reply:
Yeah, I didn't want to change the implementation from what we've traditionally used in this PR, though I would like to once we get some performance tests written.  I think a challenge is determining what implementation to use when.  For example, for a find on a short 1D uint(8) array, memchr() may be best.  For a long one, some parallel application of memchr() probably would be.  For a short array of 64-bit values, sequential search may be best; for a big one, a maxloc reduction probably would be.  Some domain maps may want (or need) to provide their own implementations for efficiency.  I worry a bit about how many implementations we'll end up wanting and how we'll choose between them, so didn't want to get into that now.